### ISSUE TITLE
Extending native types doesn't work as expected

### ISSUE THREAD

**Issue by pschanely** [pschanely/CrossHair#184]
As shown in [this example](https://crosshair-web.org/?crosshair=0.1&python=3.8&gist=07624978bddb0c5a8135ab177ed72c44), extending native types (that CrossHair cares about) doesn't work as expected.

I suspect there may be multiple layers of problems, the first of which is that a correct constructor signature is harder to deduce for these types.

cc @mristin

**Comment by mristin** [comment]
@pschanely thanks! Be aware that we used `__new__` intensively to impose constraints on primitive types in https://github.com/mristin/python-by-contract. That was a "canonical" way for me, and also the efficient one as values can not be mutated.

**Comment by pschanely** [comment]
Indeed; I recall some of that! In general, CrossHair should be able to construct via the signatures on `__new__` and `__init__` appropriately.

Even after construction, though, I expect some additional hurdles for classes extending some of the core types: str, list, dict, etc. Needs more investigation!
